vike-vue/renderer/types.ts

```diff
@@ -15,10 +15,22 @@ declare global {
       // to `false` (SPA mode).
       Page?: Component
 
-      /** Properties of the page's root Vue component. */
+      // TODO/next-major-release: remove pageProps (i.e. tell users to use data() instead of onBeforeRender() to fetch data)
+      /** Properties of the page's root Vue component - e.g. set by onBeforeRender() hook */
       pageProps?: PageProps
 
-      /** &lt;title>${title}&lt;/title> - has precedence over the config */
+      // TODO/next-major-release: remove support for setting title over onBeforeRender()
+      /** &lt;title>${title}&lt;/title> - set by onBeforeRender() hook, has precedence over the config */
+      title?: string
+
+      // Needed by getTitle()
+      data?: Data
+    }
+
+    // Can be extended by the user, see examples/basic/pages/star-wars/index/+onBeforePrerenderStart.ts
+    interface Data {
```

Comment:
@brillout the entire PR is pretty much exactly the same as https://github.com/vikejs/vike-react/pull/50 so I don't think you need to review it. (I'm pretty confident)

Except for this particular bit. I was forced to define this `Data` interface so that it can be extended in `examples/basic/pages/star-wars/index/+onBeforePrerenderStart.ts`. (Please fiddle with the code locally if unclear why.)

I couldn't come up with a better solution and I think users of `vike-react` might need the same in order to be able to implement an `onBeforePrerenderStart()` hook. Let me know if you can think of a better DX. Thanks a lot :pray:

Reply:
Makes sense. We could also mofidy Vike's `OnBeforePrerenderStartAsync` type:

```ts
type OnBeforePrerenderStartAsync<Data = unknown> = () => Promise<
  (
    | string
    | {
        url: string
        pageContext: Partial<Vike.PageContext> & { data: Data }
      }
  )[]
```

I think I've a preference for this over `interface Data` because it's more explicit. WDYT? (Feel free to disagree. I can see the 
appeal of defining `pageContext.data` globally "once and for all".)

Reply:
(To clarify a bit, the user would import the `Data` of the pre-rendered page(s) and then pass it like this `OnBeforePrerenderStartAsync<Data | Data2>`. We can release a new Vike version that adds a generic to `OnBeforePrerenderStartAsync` before merging this PR.)

Reply:
Good idea, I'll do something like that and document it. I feel it might come across as non-intuitive for the user because one part of the return type now comes from `Vike.PageContext` and another part comes from a `<Data>` generic

Reply:
https://github.com/vikejs/vike/pull/1386

Reply:
> I feel it might come across as non-intuitive for the user because one part of the return type now comes from `Vike.PageContext` and another part comes from a `<Data>` generic

I agree and I am drawn to the `inteface Data` way. I think I've a slight preference for the `OnBeforePrerenderStartAsync<Data>` way because, in real world scenarios, the vast majority of `onBeforePrerenderStart()` hooks will only provide the data of one route, sometimes two routes, and only very seldomly more than two routes. I think that's why I feel slightly more drawn to the more precise/explicit approach.

Reply:
updated :)